cheevos/cheevos.c

```diff
@@ -2902,30 +2902,23 @@ static int cheevos_iterate(coro_t *coro)
 
       MD5_Init(&coro->md5);
 
-      coro->offset = 0;
-      coro->count  = 0;
-
-      CORO_GOSUB(EVAL_MD5);
-
-      if (coro->count == 0)
+      /* Checks for the existence of a headered SNES file.
+         Unheadered files fall back to GENERIC_MD5. */
+      
+      const int SNES_HEADER_LEN = 0x200;
+      
+      if (coro->len % 0x2000 != SNES_HEADER_LEN || coro->len <= SNES_HEADER_LEN)
```

Comment:
The second clause is probably unnecessary. If len < SNES_HEADER_LEN, it won't modulus to SNES_HEADER_LEN, and if it is exactly SNES_HEADER_LEN, the file won't match using either the truncated for full file hash.

Reply:
It is better to avoid sending unnecessary requests.

Reply:
just a comment about this:

> All hashing methods are actually called for every file, regardless of core

I may be wrong, but I think the heuristic to figure the system being emulated on the cheevos context is based on the content's file extension.

And it's done only for some known SNES, Genesis and Lynx extensions. For example, SNES method will only be used if the content's file extension is one of these: 
`smc`, `fig`, `sfc`, `gd3`, `gd7`, `dx2`, `bsx`, `swc`, as we can see [in this part of the code](https://github.com/libretro/RetroArch/blob/ab2010c4940db91278ae2f0ca88c13e55e3f2065/cheevos/cheevos.c#L2594-L2605).

This snippet makes me believe that later methods will be called only if the previous ones fail:
```c
   static cheevos_finder_t finders[] =
   {
      {SNES_MD5,    "SNES (8Mb padding)",                snes_exts},
      {GENESIS_MD5, "Genesis (6Mb padding)",             genesis_exts},
      {LYNX_MD5,    "Atari Lynx (only first 512 bytes)", lynx_exts},
      {NES_MD5,     "NES (discards VROM)",               NULL},
      {GENERIC_MD5, "Generic (plain content)",           NULL},
      {FILENAME_MD5, "Generic (filename)",               NULL}
};
```

Therefore the worst case seems to be
(1) `SNES_MD5` **or** `GENESIS_MD5` **or** `LYNX_MD5` (based on the content's file extension), and then (2) `NES_MD5`, and then (3) `GENERIC_MD5`, and then (4) `FILENAME_MD5`. Which means 4 tries before total failure.

Reply:
You are right, but it is possible for someone to load a file with an extension in `snes_exts`, so why not make sure that it is worth a request before doing it? It is a cheap test.

Reply:
Updated as per my comment earlier in this thread.